rfd/0-job-worker.md

```diff
@@ -44,23 +78,152 @@ The worker library streams job output through an `io.ReadCloser`. The server wil
+
+#### Non-Considerations
+
+At its core, this is a very dangerous service. It provides a facility for remote execution and, since it needs to be privileged to set up the namespaces, to do so as root. There are a number of things that could be done to make this safer, but are not requirements of the challenge, so they will not be implemented.
+
+Some examples of security improvements in a future version would be:
+
+- dropping privileges before executing a job
+- chroot into a separate, read-only filesystem
+- better handling of commands and arguments (e.g. proper escaping)
+
 ### CLI UX
 
-### Implementation Details
+A single binary, `job-worker` is used for all actions. It has several subcommands for each role. All configuration is done via cli flags.
+
+#### root level usage output
+
+```
+A prototype job worker service that provides an api to run arbitrary linux processes
+
+Usage:
+  job-worker [command]
+
+Available Commands:
+  completion  Generate the autocompletion script for the specified shell
+  help        Help about any command
+  output      Stream the output of a job on the job-worker server
+  serve       Start the job-worker server and listen for connections
+  start       Start a job on the job-worker server
+  status      Get the status of a job on the job-worker server
+  stop        Stop a job on the job-worker server
+
+Flags:
+  -h, --help   help for job-worker
+
+Use "job-worker [command] --help" for more information about a command.
+```
+
+#### serve
+
+```
+Start the job-worker server and listen for connections
+
+Usage:
+  job-worker serve [flags]
+
+Flags:
+  -h, --help                        help for serve
+      --listen-addr string          listen address (default ":8000")
+      --max-cpu string              cpu.max value to set in cgroup for each job
+      --max-io string               io.max value to set in cgroup for each job
+      --max-memory string           memory.max value to set in cgroup for each job
+      --shutdown-timeout duration   time to wait for connections to close before forcing shutdown (default 30s)
+      --tls-ca-cert string          tls ca cert file to use for validating client certificates (required)
+      --tls-cert string             tls server certificate file (required)
+      --tls-key string              tls server key file (required)
+```
+
+#### child
+
+This is a "hidden" command that the server calls when it reexecutes itself in the new namespace before executing a job. It takes the same flags as `serve` but the additional arguments are the command and args for the job that is going to be executed.
```

Comment:
Why does it need all serve flags?

Reply:
to ensure it can reinstantiate the library with the same configuration. the important config that it needs is the cgroup limit info.

Reply:
The cgroup flags make sense, but we probably don't need the certificate-related ones.

Reply:
understood. it's easy to just pass the args that the `serve` command received to the `child` command. otherwise, we have to generate a new set of relevant flags just for the `child` command. it doesn't hurt anything to pass unnecessary flags to the `child` command, but you're right, they are unnecessary.